File: core/src/ru/deadsoftware/cavedroid/GameScreen.java
package ru.deadsoftware.cavedroid;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Screen;
import ru.deadsoftware.cavedroid.game.GameItems;
import ru.deadsoftware.cavedroid.game.GameProc;
import ru.deadsoftware.cavedroid.game.GameSaver;
import ru.deadsoftware.cavedroid.menu.MenuProc;
import ru.deadsoftware.cavedroid.misc.Assets;
import ru.deadsoftware.cavedroid.misc.InputHandlerGame;
import ru.deadsoftware.cavedroid.misc.Renderer;
import ru.deadsoftware.cavedroid.misc.states.AppState;
import ru.deadsoftware.cavedroid.misc.states.GameState;
import ru.deadsoftware.cavedroid.misc.states.MenuState;

public class GameScreen implements Screen {

    public static GameProc GP;

    public static int FPS;
    public static boolean SHOW_DEBUG = false;
    public static boolean SHOW_MAP = false;

    private Renderer renderer;
    private MenuProc menuProc;

    private InputHandlerGame inputHandlerGame;

    public GameScreen() {
        Assets.load();
        GameItems.load();
        menuProc = new MenuProc(CaveGame.TOUCH ? 320 : 480);
        renderer = menuProc;
        Gdx.input.setInputProcessor(menuProc);
    }

    public static float getWidth() {
        return Gdx.graphics.getWidth();
    }

    public static float getHeight() {
        return Gdx.graphics.getHeight();
    }

    private void game() {
        GP.update();
    }

    private void menu() {
    }

    @Override
    public void show() {
    }

    @Override
    public void render(float delta) {
        FPS = (int) (1 / delta);
        switch (CaveGame.APP_STATE) {
            case GAME:
                game();
                break;

            case MENU:
                menu();
                break;

            case LOAD:
                GP.resetRenderer();
                renderer = GP.renderer;
                if (inputHandlerGame == null) {
                    inputHandlerGame = new InputHandlerGame();
                }
                Gdx.input.setInputProcessor(inputHandlerGame);
                CaveGame.APP_STATE = AppState.GAME;
                CaveGame.GAME_STATE = GameState.PLAY;
                break;

            case SAVE:
                GameSaver.save(GP);
                CaveGame.APP_STATE = AppState.MENU;
                CaveGame.MENU_STATE = MenuState.MAIN;
                GP.dispose();
                renderer = menuProc;
                Gdx.input.setInputProcessor(menuProc);
                break;
        }
        renderer.render();
    }

    @Override
    public void resize(int width, int height) {
        switch (CaveGame.APP_STATE) {
            case MENU:
                menuProc = new MenuProc(CaveGame.TOUCH ? 320 : 480);
                Gdx.input.setInputProcessor(menuProc);
                renderer = menuProc;
                break;
            case GAME:
                GP.resetRenderer();
                renderer = GP.renderer;
                break;
        }
    }

    @Override
    public void pause() {

    }

    @Override
    public void resume() {

    }

    @Override
    public void hide() {

    }

    @Override
    public void dispose() {

    }

}

File: core/gdx/src/main/kotlin/ru/fredboy/cavedroid/gdx/game/GameScreen.kt
package ru.fredboy.cavedroid.gdx.game

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.math.Rectangle
import ru.fredboy.cavedroid.common.model.Joystick
import ru.fredboy.cavedroid.common.model.StartGameConfig
import ru.fredboy.cavedroid.data.configuration.model.GameContext
import ru.fredboy.cavedroid.domain.configuration.model.CameraContext
import ru.fredboy.cavedroid.domain.configuration.repository.ApplicationContextRepository
import ru.fredboy.cavedroid.domain.items.repository.MobParamsRepository
import ru.fredboy.cavedroid.gdx.base.BaseScreen
import ru.fredboy.cavedroid.gdx.game.di.DaggerGameComponent
import ru.fredboy.cavedroid.gdx.game.di.GameComponent
import ru.fredboy.cavedroid.gdx.utils.applicationComponent
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class GameScreen @Inject constructor(
    private val applicationContextRepository: ApplicationContextRepository,
    private val mobParamsRepository: MobParamsRepository,
) : BaseScreen(applicationContextRepository) {

    override val scaleFactor: Float
        get() = 0.5f

    var gameComponent: GameComponent? = null
        private set

    private fun getGameContext(gameConfig: StartGameConfig): GameContext = GameContext(
        isLoadGame = gameConfig is StartGameConfig.Load,
        saveGameDirectory = gameConfig.saveDirectory,
        worldName = gameConfig.worldName,
        showInfo = false,
        showMap = false,
        joystick = Joystick(requireNotNull(mobParamsRepository.getMobParamsByKey("char")).speed),
        cameraContext = CameraContext(
            viewport = Rectangle(
                /* x = */ 0f,
                /* y = */ 0f,
                /* width = */ applicationContextRepository.getWidth(),
                /* height = */ applicationContextRepository.getHeight(),
            ),
            visibleWorld = Rectangle(
                /* x = */ 0f,
                /* y = */ 0f,
                /* width = */ applicationContextRepository.getWidth(),
                /* height = */ applicationContextRepository.getHeight(),
            ),
        ),
    )

    private fun getGameComponent(gameConfig: StartGameConfig): GameComponent {
        val gameContext = getGameContext(gameConfig)

        return DaggerGameComponent.builder()
            .applicationComponent(Gdx.app.applicationListener.applicationComponent)
            .gameContext(gameContext)
            .build()
    }

    private fun resetGameComponent() {
        gameComponent?.gameProc?.dispose()
        gameComponent = null
    }

    fun newGame(gameConfig: StartGameConfig.New) {
        resetGameComponent()
        gameComponent = getGameComponent(gameConfig).apply {
            gameProc.setPlayerGameMode(gameConfig.gameMode)
            gameRenderer.render(0f)
            gameSaveHelper.saveGame(false)
        }
    }

    fun loadGame(gameConfig: StartGameConfig.Load) {
        resetGameComponent()
        gameComponent = getGameComponent(gameConfig)
    }

    fun saveGame() {
        val gameComponent = requireNotNull(gameComponent) {
            "GameScreen#saveGame: gameComponent was not set before saveGame()"
        }

        gameComponent.gameRenderer.render(0f)
        gameComponent.gameSaveHelper.saveGame(overwrite = true)
    }

    override fun show() {
        val gameComponent = requireNotNull(gameComponent) {
            "GameScreen#show: gameComponent was not set before show"
        }
        gameComponent.gameProc.show()
        gameComponent.gameRenderer.render(0f)
    }

    override fun render(delta: Float) {
        val proc = requireNotNull(gameComponent?.gameProc) {
            "GameScreen#render: gameComponent was not set before render"
        }
        proc.update(delta)
    }

    override fun onResize(width: Int, height: Int) {
        gameComponent?.gameContextRepository?.getCameraContext()?.let { cameraContext ->
            cameraContext.viewport.apply {
                setWidth(applicationContextRepository.getWidth())
                setHeight(applicationContextRepository.getHeight())
            }

            cameraContext.visibleWorld.apply {
                setWidth(applicationContextRepository.getWidth())
                setHeight(applicationContextRepository.getHeight())
            }
        }

        gameComponent?.gameProc?.onResize()
    }

    override fun pause() {
        gameComponent?.gameProc?.onGamePaused()
    }

    override fun resume() {
        gameComponent?.gameProc?.onGameResumed()
    }

    override fun hide() {
    }

    override fun onDispose() {
        resetGameComponent()
    }
}
